### ISSUE TITLE
NPE on CompositionLayer

### ISSUE THREAD

**Issue by stefanomemrise** [airbnb/lottie-android#524]
Hi!
This has been happening recently, last release included:

Fatal Exception: java.lang.NullPointerException
       at com.airbnb.lottie.model.layer.CompositionLayer.<init>(SourceFile:8120)
       at com.airbnb.lottie.LottieDrawable.b(SourceFile:201)
       at com.airbnb.lottie.LottieAnimationView.setComposition(SourceFile:11164)

**Comment by gpeal** [comment]
@stefanomemrise Can you include a source mapped stack trace? I can't debug anything without a repro or a stack trace.

**Comment by stefanomemrise** [comment]
Fatal Exception: java.lang.NullPointerException
       at com.airbnb.lottie.model.layer.CompositionLayer.<init>(SourceFile:8120)
       at com.airbnb.lottie.LottieDrawable.b(SourceFile:201)
       at com.airbnb.lottie.LottieAnimationView.setComposition(SourceFile:11164)
       at com.memrise.android.memrisecompanion.ui.presenter.view.FlowerView$$Lambda$0.a(Unknown Source:1029)
       at com.memrise.android.memrisecompanion.ui.util.WordDrawableMapper$$Lambda$0.a(Unknown Source:1053)
       at com.airbnb.lottie.model.FileCompositionLoader.onPostExecute(SourceFile:1024)
       at android.os.AsyncTask.finish(AsyncTask.java:632)
       at android.os.AsyncTask.access$600(AsyncTask.java:177)
       at android.os.AsyncTask$InternalHandler.handleMessage(AsyncTask.java:645)
       at android.os.Handler.dispatchMessage(Handler.java:110)
       at android.os.Looper.loop(Looper.java:193)
       at android.app.ActivityThread.main(ActivityThread.java:5341)
       at java.lang.reflect.Method.invokeNative(Method.java)
       at java.lang.reflect.Method.invoke(Method.java:515)
       at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:830)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:646)
       at dalvik.system.NativeStart.main(NativeStart.java)

**Comment by stefanomemrise** [comment]
Also, it's only happening on Android 4.x

**Comment by gpeal** [comment]
@stefanomemrise This still isn't source mapped. The line numbers aren't correct.

**Comment by yangheng09103** [comment]
Fatal Exception: java.lang.NullPointerException
com.airbnb.lottie.model.layer.CompositionLayer.<init> (CompositionLayer.java:70)
--
  | com.airbnb.lottie.LottieDrawable.buildCompositionLayer (LottieDrawable.java:195)
  | com.airbnb.lottie.LottieDrawable.setComposition (LottieDrawable.java:156)
  | com.airbnb.lottie.LottieAnimationView.setComposition (LottieAnimationView.java:461)
  | com.airbnb.lottie.LottieAnimationView$3.onCompositionLoaded (LottieAnimationView.java:426)
  | com.airbnb.lottie.model.FileCompositionLoader.onPostExecute (FileCompositionLoader.java:24)
  | com.airbnb.lottie.model.FileCompositionLoader.onPostExecute (FileCompositionLoader.java:10)

**Comment by gpeal** [comment]
@yangheng09103 @stefanomemrise Can you attach the animation that crashes? Does it crash 100% of the time?

@yangheng09103 does yours only crash on 4.x like @stefanomemrise?

**Comment by yangheng09103** [comment]
It's not 100% of the time.It's only happened on 4.x.I got this crash from fabric after lottie updated to v2.3.0.This crash is new since 2.3.0.

**Comment by yangheng09103** [comment]
[anim_save.txt](https://github.com/airbnb/lottie-android/files/1503984/anim_save.txt)
This is the animation file.

**Comment by magic1983** [comment]
#0. Crashed: main: 0 0 0x0000000000000000
       at android.view.LayoutInflater.createView(LayoutInflater.java:620)
       at android.view.LayoutInflater.createViewFromTag(LayoutInflater.java:696)
       at android.view.LayoutInflater.rInflate(LayoutInflater.java:755)
       at android.view.LayoutInflater.inflate(LayoutInflater.java:492)
       at android.view.LayoutInflater.inflate(LayoutInflater.java:397)
       at android.view.LayoutInflater.inflate(LayoutInflater.java:353)
       at mobi.trustlab.boost.view.floatingview.BoostFloatingView.(BoostFloatingView.kt)
       at mobi.trustlab.boost.main.boost.ClearBoostCard.clearBoost(ClearBoostCard.kt:207)
       at mobi.trustlab.boost.main.boost.ClearBoostCard$2.onClick(ClearBoostCard.kt:60)
       at android.view.View.performClick(View.java:4597)
       at android.view.View$PerformClick.run(View.java:18907)
       at android.os.Handler.handleCallback(Handler.java:808)
       at android.os.Handler.dispatchMessage(Handler.java:103)
       at android.os.Looper.loop(Looper.java:193)
       at android.app.ActivityThread.main(ActivityThread.java:5398)
       at java.lang.reflect.Method.invokeNative(Method.java)
       at java.lang.reflect.Method.invoke(Method.java:515)
       at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:940)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:756)
       at dalvik.system.NativeStart.main(NativeStart.java)

--

Fatal Exception: android.view.InflateException: Binary XML file line #9: Error inflating class <unknown>
       at android.view.LayoutInflater.createView(LayoutInflater.java:620)
       at android.view.LayoutInflater.createViewFromTag(LayoutInflater.java:696)
       at android.view.LayoutInflater.rInflate(LayoutInflater.java:755)
       at android.view.LayoutInflater.inflate(LayoutInflater.java:492)
       at android.view.LayoutInflater.inflate(LayoutInflater.java:397)
       at android.view.LayoutInflater.inflate(LayoutInflater.java:353)
       at mobi.trustlab.boost.view.floatingview.BoostFloatingView.(BoostFloatingView.kt)
       at mobi.trustlab.boost.main.boost.ClearBoostCard.clearBoost(ClearBoostCard.kt:207)
       at mobi.trustlab.boost.main.boost.ClearBoostCard$2.onClick(ClearBoostCard.kt:60)
       at android.view.View.performClick(View.java:4597)
       at android.view.View$PerformClick.run(View.java:18907)
       at android.os.Handler.handleCallback(Handler.java:808)
       at android.os.Handler.dispatchMessage(Handler.java:103)
       at android.os.Looper.loop(Looper.java:193)
       at android.app.ActivityThread.main(ActivityThread.java:5398)
       at java.lang.reflect.Method.invokeNative(Method.java)
       at java.lang.reflect.Method.invoke(Method.java:515)
       at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:940)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:756)
       at dalvik.system.NativeStart.main(NativeStart.java)
Caused by java.lang.reflect.InvocationTargetException
       at java.lang.reflect.Constructor.constructNative(Constructor.java)
       at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
       at android.view.LayoutInflater.createView(LayoutInflater.java:594)
       at android.view.LayoutInflater.createViewFromTag(LayoutInflater.java:696)
       at android.view.LayoutInflater.rInflate(LayoutInflater.java:755)
       at android.view.LayoutInflater.inflate(LayoutInflater.java:492)
       at android.view.LayoutInflater.inflate(LayoutInflater.java:397)
       at android.view.LayoutInflater.inflate(LayoutInflater.java:353)
       at mobi.trustlab.boost.view.floatingview.BoostFloatingView.(BoostFloatingView.kt)
       at mobi.trustlab.boost.main.boost.ClearBoostCard.clearBoost(ClearBoostCard.kt:207)
       at mobi.trustlab.boost.main.boost.ClearBoostCard$2.onClick(ClearBoostCard.kt:60)
       at android.view.View.performClick(View.java:4597)
       at android.view.View$PerformClick.run(View.java:18907)
       at android.os.Handler.handleCallback(Handler.java:808)
       at android.os.Handler.dispatchMessage(Handler.java:103)
       at android.os.Looper.loop(Looper.java:193)
       at android.app.ActivityThread.main(ActivityThread.java:5398)
       at java.lang.reflect.Method.invokeNative(Method.java)
       at java.lang.reflect.Method.invoke(Method.java:515)
       at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:940)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:756)
       at dalvik.system.NativeStart.main(NativeStart.java)
Caused by java.lang.NullPointerException
       at com.airbnb.lottie.model.layer.CompositionLayer.(CompositionLayer.java)
       at com.airbnb.lottie.LottieDrawable.buildCompositionLayer(LottieDrawable.java:201)
       at com.airbnb.lottie.LottieDrawable.enableMergePathsForKitKatAndAbove(LottieDrawable.java:113)
       at com.airbnb.lottie.LottieAnimationView.enableMergePathsForKitKatAndAbove(LottieAnimationView.java:263)
       at com.airbnb.lottie.LottieAnimationView.init(LottieAnimationView.java:114)
       at com.airbnb.lottie.LottieAnimationView.(LottieAnimationView.java)
       at java.lang.reflect.Constructor.constructNative(Constructor.java)
       at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
       at android.view.LayoutInflater.createView(LayoutInflater.java:594)
       at android.view.LayoutInflater.createViewFromTag(LayoutInflater.java:696)
       at android.view.LayoutInflater.rInflate(LayoutInflater.java:755)
       at android.view.LayoutInflater.inflate(LayoutInflater.java:492)
       at android.view.LayoutInflater.inflate(LayoutInflater.java:397)
       at android.view.LayoutInflater.inflate(LayoutInflater.java:353)
       at mobi.trustlab.boost.view.floatingview.BoostFloatingView.(BoostFloatingView.kt)
       at mobi.trustlab.boost.main.boost.ClearBoostCard.clearBoost(ClearBoostCard.kt:207)
       at mobi.trustlab.boost.main.boost.ClearBoostCard$2.onClick(ClearBoostCard.kt:60)
       at android.view.View.performClick(View.java:4597)
       at android.view.View$PerformClick.run(View.java:18907)
       at android.os.Handler.handleCallback(Handler.java:808)
       at android.os.Handler.dispatchMessage(Handler.java:103)
       at android.os.Looper.loop(Looper.java:193)
       at android.app.ActivityThread.main(ActivityThread.java:5398)
       at java.lang.reflect.Method.invokeNative(Method.java)
       at java.lang.reflect.Method.invoke(Method.java:515)
       at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:940)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:756)
       at dalvik.system.NativeStart.main(NativeStart.java)

**Comment by gpeal** [comment]
@stefanomemrise @yangheng09103 @magic1983 
I can't repro this. It would be really helpful to have a sample project or even a link number for the actual crash `at com.airbnb.lottie.model.layer.CompositionLayer.(CompositionLayer.java)`

**Comment by stefanomemrise** [comment]
In our case it looks like it was related to the D8 compiler of gradle. After switching back to DX it never happened again. Hope this helps.

**Comment by stefanomemrise** [comment]
In the gradle.properties:
android.enableD8=false

**Comment by yangheng09103** [comment]
I find out that code revision 3867cdd is match the crash trace from the Fabric 
Crashlytics log.
`Fatal Exception: java.lang.NullPointerException
com.airbnb.lottie.model.layer.CompositionLayer. (CompositionLayer.java:70)
  | com.airbnb.lottie.LottieDrawable.buildCompositionLayer (LottieDrawable.java:195)
  | com.airbnb.lottie.LottieDrawable.setComposition (LottieDrawable.java:156)
  | com.airbnb.lottie.LottieAnimationView.setComposition (LottieAnimationView.java:461)
  | com.airbnb.lottie.LottieAnimationView$3.onCompositionLoaded (LottieAnimationView.java:426)
  | com.airbnb.lottie.model.FileCompositionLoader.onPostExecute (FileCompositionLoader.java:24)
  | com.airbnb.lottie.model.FileCompositionLoader.onPostExecute (FileCompositionLoader.java:10)`

The code crash should be 
`BaseLayer parentLayer = layerMap.get(layerView.getLayerModel().getParentId());`
It looks like `layerView.getLayerModel()` was null at some time.

**Comment by gpeal** [comment]
@yangheng09103 @magic1983 Are you using D8 as well?

**Comment by mhdtouban** [comment]
I am using D8 #544

**Comment by yangheng09103** [comment]
Yes,D8 is enabled in my project.

**Comment by gpeal** [comment]
@yangheng09103 @mhdtouban @stefanomemrise Are there any trends in the devices themselves beyond just being 4.x?

**Comment by stefanomemrise** [comment]
most of the models are based on Cortex-A7 - and for sure 90% of them on Android 4.x

**Comment by gpeal** [comment]
@stefanomemrise are there some on 5.x+?

**Comment by stefanomemrise** [comment]
96% on 4, 1% on 5/6/7

**Comment by Bridouille** [comment]
Hi! I'm also getting this errors in my FirebaseCrashReporting logs.

```
Exception java.lang.NullPointerException: layerView is null!
com.airbnb.lottie.model.layer.CompositionLayer.addValueCallback (CompositionLayer.java)
com.airbnb.lottie.model.layer.CompositionLayer.<init> (CompositionLayer.java)
com.airbnb.lottie.model.layer.BaseLayer.access$000 (BaseLayer.java)
com.airbnb.lottie.model.layer.CompositionLayer.<init> (CompositionLayer.java)
com.airbnb.lottie.LottieDrawable.buildCompositionLayer (LottieDrawable.java)
com.airbnb.lottie.LottieDrawable.access$000 (LottieDrawable.java)
com.airbnb.lottie.LottieAnimationView.setComposition (LottieAnimationView.java)
com.airbnb.lottie.LottieAnimationView$3.onCompositionLoaded (LottieAnimationView.java)
com.airbnb.lottie.model.FileCompositionLoader.doInBackground (FileCompositionLoader.java)
com.airbnb.lottie.model.FileCompositionLoader.onPostExecute (FileCompositionLoader.java)
android.os.AsyncTask.finish (AsyncTask.java:632)
android.os.AsyncTask.access$600 (AsyncTask.java:177)
android.os.AsyncTask$InternalHandler.handleMessage (AsyncTask.java:645)
android.os.Handler.dispatchMessage (Handler.java:110)
android.os.Looper.loop (Looper.java:193)
android.app.ActivityThread.main (ActivityThread.java:5299)
java.lang.reflect.Method.invokeNative (Method.java)
java.lang.reflect.Method.invoke (Method.java:515)
com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run (ZygoteInit.java:829)
com.android.internal.os.ZygoteInit.main (ZygoteInit.java:645)
dalvik.system.NativeStart.main (NativeStart.java)
```

It only happens on devices running API 19 (4.4), not all the time apparently and I don't know for which annimation in perticular as this is the only stacktrace I have (happens on various screens).

Previously I was setting the `app:lottie_fileName ` in XML and got the error when the whole view inflates, now I moved the `setAnimation()` in code at runtime with a try catch block to maybe intercept this exception and simply not show the animation, but looks like it doesn't work either.

**Comment by gpeal** [comment]
@yangheng09103 yours is https://github.com/airbnb/lottie-android/issues/590

**Comment by mhdtouban** [comment]
Did you solve this issue? it's still being reported on v2.3.1
Android 4.4.2
smartphone HTC Desire 820G PLUS dual sim

**Comment by gpeal** [comment]
@mhdtouban The fix is in 2.5.0-rc1. rc2 will come out this week.
